### ISSUE TITLE
NuGet on Mono workarounds

### ISSUE THREAD

**Issue by mrward** [NuGet/Home#2992]
Xamarin Studio 6.1 now ships with NuGet 3.4.3 and is using a custom build of the NuGet binaries in order to workaround the following two Mono problems:
1. [InvalidOperationException when running NuGet](https://bugzilla.xamarin.com/show_bug.cgi?id=37970)
   1. [NuGet 3 patch](https://github.com/mrward/NuGet.Client/commit/f2362284774db37f3b868a6c704aec21c391056c)
2. [WebException.Response is null when https request needs proxy authentication](https://bugzilla.xamarin.com/show_bug.cgi?id=19594)
   1. [NuGet 3 patch](https://github.com/mrward/NuGet.Client/commit/a0e7ae780215c2b2a87816974c22b474eb9257d5)

Ideally these bugs would be fixed in Mono but for now I have worked around them in NuGet.

Would it make sense for these patches to be included here in the main NuGet source code? If so I can update the code to work with the dev branch and open a pull request for each workaround.

**Comment by yishaigalatzer** [comment]
Can we accelerate the mono fixes instead of integrating ugly hacks into the NuGet codebase (that then never go away?).

I don't quite understand the release cadence for mono, so I'll start a thread to figure this out.

**Comment by mrward** [comment]
The InvalidOperationException bug has been fixed in Mono 4.5.2.360 (master/d1d72cd Mon Jun 20 09:32:05 EDT 2016).

**Comment by yishaigalatzer** [comment]
Thanks @mrward , we are waiting on the decision on the http client issue from your side, and than we can proceed from here. I'm pulling this into the 3.5 release so it doesn't easily fall off the radar

**Comment by jzeferino** [comment]
@yishaigalatzer , @mrward,

 Using Mono 4.4.1 and nuget 3.4.4.1321.
When trying to create a simple nuget package:

 **nuget pack cpv.nuspec -verbosity detailed**

> Attempting to build package from 'cpv.nuspec'.
> System.InvalidOperationException: Cannot create a package that has no dependencies nor content.
>   at NuGet.Packaging.PackageBuilder.Save (System.IO.Stream stream) <0x34e61c8 + 0x002f3> in <filename unknown>:0 
>   at NuGet.CommandLine.PackCommand.BuildPackage (NuGet.Packaging.PackageBuilder builder, System.String outputPath) <0x34e4af8 + 0x001ef> in <filename unknown>:0 

This is the nuspec file:

```
<package>
  <metadata>
    <id>cpv</id>
    <version>0.0.1</version>
    <title>cpv</title>
    <authors>xxx</authors>
    <owners>xxx</owners>
    <licenseUrl>https://github.com/xxx/blob/master/LICENSE</licenseUrl>
    <projectUrl>https://github.com/xxx</projectUrl>
    <requireLicenseAcceptance>true</requireLicenseAcceptance>
    <description>cpv test</description>
  </metadata>
  <files>
    <file src="../cpv/cpv/bin/Release/cpv.dll" target="lib/MonoAndroid403" />  
  </files>
</package>

```

Any fix?

**Comment by mrward** [comment]
I believe this is because of a problem handling the relative path. Possible things to try:
1. Specify BasePath in the NuGet.exe command line.
   1. `mono NuGet.exe pack YourNuSpec.nuspec -BasePath ./`
2. Do not use `../` in your file path. So copy the .dll so it is inside the directory where your .nuspec is.
   1. `<file src="cpv/cpv/bin/Release/cpv.dll" target="lib/MonoAndroid403" />`

**Comment by jzeferino** [comment]
Thank you @mrward .

`mono NuGet.exe pack YourNuSpec.nuspec -BasePath ./` worked.
and removing the `../` and palcing the nuspec in that path worked too.

Personaly I prefer to have the `nuspec` folder at the same level of the folder with the `src` so in this case I will pick the first solution. Not so clean but, I will wrapped it in a `.sh` and it will be fine.

**Comment by rohit21agrawal** [comment]
@mrward can you provide the steps to repro this issue from NuGet running on Mono ? Will need them to do some checking (and possibly write tests) while integrating your your HttpRequest patch.

**Comment by mrward** [comment]
@rohit21agrawal - For the mono proxy patch I usually test this with [Fiddler](http://www.telerik.com/fiddler) installed on Windows and use it as a reverse proxy for the Mac.

**To enable the reverse proxy:**

http://docs.telerik.com/fiddler/configure-fiddler/tasks/usefiddlerasreverseproxy

**To enable proxy authentication in Fiddler:**

Select Rules menu, then select Require Proxy Authentication.

**To setup the Mac to use a proxy:**
1. System Preferences - Network - Advanced - Proxies.
2. Tick both:
   1. "Web Proxy (HTTP)"
   2.  "Secure Web Proxy (HTTPS)"
3. Enter the IP address of your Windows machine and the Fiddler port 8888.
4. Click OK.
5. Click Apply.

Using unpatched NuGet on Mono and trying to connect to https://www.nuget.org/api/v2 will fail with an error similar to the following:

`System.Net.WebException: Error getting response stream (The remote server returned a 407 status code.): SecureChannelFailure ---> System.Exception:    at System.Net.WebConnection.HandleError(WebExceptionStatus st, System.Exception e, System.String where)`.

Connecting to http://www.nuget.org/api/v2 - so not using https - will work OK if proxy credentials are available.

This is tested with Xamarin Studio using NuGet.Core.dll or the NuGet v3 assemblies.

**Comment by rohit21agrawal** [comment]
@mrward i am using mono version 4.4.1, and even without doing any proxy related stuff, if i do something like:

mono nuget.exe list -source http://www.nuget.org/api/v2

it says : ProxyNameResolutionFailure

any idea what's going on here?

**Comment by mrward** [comment]
@rohit21agrawal No idea. It works fine for me by default with no proxy but I have no system wide proxy and I am not on a company network.

I also tried out some other tests with Mono 4.4.1 and NuGet.exe that ships with Mono, which is currently 2.8.5, and also NuGet 3.4.4.

**Proxy enabled on Mac - no proxy configured for NuGet - NuGet 2.8.5**

```
nuget list -source http://www.nuget.org/api/v2
Error: ConnectFailure (Connection refused)
```

**Proxy enabled on Mac - no proxy configured for NuGet - NuGet 3.4.4**

```
 mono nuget-3.4.4.exe list -source http://www.nuget.org/api/v2
 Error: ConnectFailure (Connection refused)
 Connection refused
```

**Enabling proxy**

```
nuget config -set http_proxy=http://proxy-ip-address:8888
```

(replacing proxy-ip-address above with the real value)

**Proxy enabled on Mac - proxy configured for NuGet - NuGet 2.8.5**

```
nuget list -source http://www.nuget.org/api/v2
Please provide proxy credentials:

nuget list -source https://www.nuget.org/api/v2
Error getting response stream (The remote server returned a 407 status code.):    SecureChannelFailure
```

**Proxy enabled on Mac - proxy configured for NuGet - NuGet 3.4.4**

```
mono nuget-3.4.4.exe list -source http://www.nuget.org/api/v2
Please provide proxy credentials:

mono nuget-3.4.4.exe list -source https://www.nuget.org/api/v2
Error getting response stream (The remote server returned a 407 status code.):     SecureChannelFailure
    at System.Net.WebConnection.HandleError (WebExceptionStatus st, System.Exception e, System.String where) [0x00000] in <filename unknown>:0 
    at System.Net.WebConnection.CreateTunnel (System.Net.HttpWebRequest request, System.Uri connectUri, System.IO.Stream stream, System.Byte[]& buffer) [0x00000] in <filename unknown>:0 
    at System.Net.WebConnection.CreateStream (System.Net.HttpWebRequest request) [0x00000] in <filename unknown>:0 
    at System.Net.WebConnection.InitConnection (System.Object state) [0x00000] in <filename unknown>:0 
    at System.Net.WebConnection.<WebConnection>m__0 (System.Object state) [0x00000] in <filename unknown>:0 
    at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context (System.Object state) [0x00000] in <filename unknown>:0 
    at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, Boolean preserveSyncCtx) [0x00000] in <filename unknown>:0 
    at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, Boolean preserveSyncCtx) [0x00000] in <filename unknown>:0 
    at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () [0x00000] in <filename unknown>:0 
    at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00000] in <filename unknown>:0 
    at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00000] in <filename unknown>:0
```

**Comment by rohit21agrawal** [comment]
@mrward i have a pull request out integrating your patch into the dev branch. would be great if you can help validate as well.

**Comment by mrward** [comment]
@rohit21agrawal - That NuGet.exe does not seem to fix the problem. I still see the same error for the https url.

```
mono NuGet-rohit.exe
NuGet Version: 3.5.0.1367

mono NuGet-rohit.exe list -source https://www.nuget.org/api/v2
Error getting response stream (The remote server returned a 407 status code.): SecureChannelFailure
    at System.Net.WebConnection.HandleError (WebExceptionStatus st, System.Exception e, System.String where) [0x00000] in <filename unknown>:0 
```

**Comment by rohit21agrawal** [comment]
@mrward just realized that nuget list command calls into the v2 codebase, and we haven't made the fix there.
Can you try a command like install?

**Comment by mrward** [comment]
Yeah that works correctly. I get a proxy credentials prompt when using the https source.

```
mono NuGet-rohit.exe install NUnit -source https://www.nuget.org/api/v2
Feeds used:
  https://www.nuget.org/api/v2

Please provide proxy credentials:
```

Trying the same thing with NuGet 3.4.4 I see:

```
Unable to load the service index for source https://www.nuget.org/api/v2.
  Error getting response stream (The remote server returned a 407 status code.):     SecureChannelFailure
    at System.Net.WebConnection.HandleError (WebExceptionStatus st, Syst
```

**Comment by mrward** [comment]
Update: The workaround added for proxy auth failing with https on Mono no longer seems to work with Mono 4.6, nor Mono 4.8, nor latest from mono/master. So there is never any proxy auth prompt with these versions of Mono. Mono 4.4 still seems to be OK.

**Comment by mrward** [comment]
Update: The [WebException.Response is null when https request needs proxy authentication](https://bugzilla.xamarin.com/show_bug.cgi?id=19594) problem has been fixed in Mono 4.8.0. This fixes NuGet v2 on Mono not prompting for proxy credentials.

There is a problem NuGet v3 not prompting for proxy credentials with Mono. I have opened a separate bug report for this other problem:

[HttpClientHandler.SendAsync should throw HttpRequestException for proxy auth failure](https://bugzilla.xamarin.com/show_bug.cgi?id=44978)

**Comment by mrward** [comment]
- [Bug 44978 - HttpClientHandler.SendAsync should throw HttpRequestException for proxy auth failure](https://bugzilla.xamarin.com/show_bug.cgi?id=44978) has been fixed in Mono 4.8.0.

Further testing has revealed another Mono problem with proxy credentials not being used for https package sources:
- [Bug 45108 - Proxy credentials not used for https url](https://bugzilla.xamarin.com/show_bug.cgi?id=45108)

**Comment by mrward** [comment]
- [Bug 45108 - Proxy credentials not used for https url](https://bugzilla.xamarin.com/show_bug.cgi?id=45108) has been fixed in Mono 4.8.0.
